File: the-simple-graphql-service/python/the_simple_graphql_service/the_simple_graphql_service_stack.py
from aws_cdk import (
    aws_lambda as _lambda,
    aws_appsync as appsync,
    aws_dynamodb as dynamo_db,
    core
)
import os


class TheSimpleGraphqlServiceStack(core.Stack):
    def __init__(self, scope: core.Construct, id: str, **kwargs) -> None:
        super().__init__(scope, id, **kwargs)
        
        schema_location = os.path.dirname(os.path.realpath(__file__)) + "/../schema/schema.graphql"

        # Create a new AppSync GraphQL API
        api = appsync.GraphqlApi(self, 'Api',
                                 name="demoapi",
                                 log_config=appsync.LogConfig(field_log_level=appsync.FieldLogLevel.ALL),
                                 schema=appsync.Schema.from_asset(schema_location)
                                 )

        api_key = appsync.CfnApiKey(self, 'the-simple-graphql-service-api-key',
                                    api_id=api.api_id
                                    )

        # Create new DynamoDB Table for Customer
        customer_table = dynamo_db.Table(self, "CustomerTable",
                                         partition_key=dynamo_db.Attribute(name="id",
                                                                           type=dynamo_db.AttributeType.STRING)
                                         )

        # Add Customer DynamoDB as a Datasource for the Graphql API.
        customer_ds = api.add_dynamo_db_data_source('Customer', customer_table)

        # Query Resolver to get all Customers
        customer_ds.create_resolver(
            type_name='Query',
            field_name='getCustomers',
            request_mapping_template=appsync.MappingTemplate.dynamo_db_scan_table(),
            response_mapping_template=appsync.MappingTemplate.dynamo_db_result_list(),
        )

        # Query Resolver to get an individual Customer by their id
        customer_ds.create_resolver(
            type_name='Query',
            field_name='getCustomer',
            request_mapping_template=appsync.MappingTemplate.dynamo_db_get_item('id', 'id'),
            response_mapping_template=appsync.MappingTemplate.dynamo_db_result_item(),
        )

        # Mutation Resolver for adding a new Customer
        customer_ds.create_resolver(
            type_name='Mutation',
            field_name='addCustomer',
            request_mapping_template=appsync.MappingTemplate.dynamo_db_put_item(
                key=appsync.PrimaryKey.partition('id').auto(),
                values=appsync.Values.projecting('customer')
            ),
            response_mapping_template=appsync.MappingTemplate.dynamo_db_result_item()
        )

        # Mutation Resolver for updating an existing Customer
        customer_ds.create_resolver(
            type_name='Mutation',
            field_name='saveCustomer',
            request_mapping_template=appsync.MappingTemplate.dynamo_db_put_item(
                key=appsync.PrimaryKey.partition('id').is_('id'),
                values=appsync.Values.projecting('customer')
            ),
            response_mapping_template=appsync.MappingTemplate.dynamo_db_result_item()
        )

        #  Mutation resolver for creating a new customer along with their first order 
        customer_ds.create_resolver(
            type_name='Mutation',
            field_name='saveCustomerWithFirstOrder',
            request_mapping_template=appsync.MappingTemplate.dynamo_db_put_item(
                key=appsync.PrimaryKey.partition('order').auto().sort('customer').is_('customer.id'),
                values=appsync.Values.projecting('order').attribute('referral').is_('referral')
            ),
            response_mapping_template=appsync.MappingTemplate.dynamo_db_result_item()
        )

        # Mutation Resolver for deleting an existing Customer
        customer_ds.create_resolver(
            type_name='Mutation',
            field_name='removeCustomer',
            request_mapping_template=appsync.MappingTemplate.dynamo_db_delete_item('id', 'id'),
            response_mapping_template=appsync.MappingTemplate.dynamo_db_result_item(),
        )

        # defines an AWS  Lambda resource
        loyalty_lambda = _lambda.Function(self, "LoyaltyLambdaHandler",
                                          runtime=_lambda.Runtime.NODEJS_12_X,
                                          handler="loyalty.handler",
                                          code=_lambda.Code.from_asset("lambda_fns"),
                                          )

        # Add Loyalty Lambda as a Datasource for the Graphql API.
        loyalty_ds = api.add_lambda_data_source('Loyalty', loyalty_lambda)

        # Query Resolver to get all Customers
        loyalty_ds.create_resolver(
            type_name='Query',
            field_name='getLoyaltyLevel',
            request_mapping_template=appsync.MappingTemplate.lambda_request(),
            response_mapping_template=appsync.MappingTemplate.lambda_result(),
        )

        # GraphQL API Endpoint
        core.CfnOutput(self, 'Endpoint',
                       value=api.graphql_url
                       )

        # API Key
        core.CfnOutput(self, 'API_Key',
                       value=api_key.attr_api_key
                       )

File: the-simple-graphql-service/csharp/src/TheSimpleGraphqlService/TheSimpleGraphqlServiceStack.cs
using Amazon.CDK;
using Lambda = Amazon.CDK.AWS.Lambda;
using DynamoDB = Amazon.CDK.AWS.DynamoDB;
using AppSync = Amazon.CDK.AWS.AppSync;

namespace TheSimpleGraphqlService
{
    public class TheSimpleGraphqlServiceStack : Stack
    {

        private readonly AppSync.GraphqlApi _graphqlApi;
        private readonly AppSync.CfnApiKey _graphqlKey;
        private readonly DynamoDB.Table _customerTable;
        private readonly Lambda.Function _loyaltyLambda;

        internal TheSimpleGraphqlServiceStack(Construct scope, string id, IStackProps props = null) : base(scope, id, props)
        {
            /**
            * Create a new AppSync GraphQL API
            */
            _graphqlApi = new AppSync.GraphqlApi(this, "Api", new AppSync.GraphqlApiProps
            {
                Name = "demoapi",
                LogConfig = new AppSync.LogConfig
                {
                    FieldLogLevel = AppSync.FieldLogLevel.ALL
                },
                Schema = AppSync.Schema.FromAsset("./schema/schema.graphql")
            });

            /**
             * Create Appsync Api Key
             */
            _graphqlKey = new AppSync.CfnApiKey(this, "the-simple-graphql-service-api-key", new AppSync.CfnApiKeyProps
            {
                ApiId = _graphqlApi.ApiId
            });


            /**
             * Create new DynamoDB Table for Customer
             */
            _customerTable = new DynamoDB.Table(this, "CustomerTable", new DynamoDB.TableProps
            {
                BillingMode = DynamoDB.BillingMode.PAY_PER_REQUEST,
                PartitionKey = new DynamoDB.Attribute
                {
                    Name = "id",
                    Type = DynamoDB.AttributeType.STRING
                }
            });

            /**
             * Add Customer DynamoDB as a Datasource for the Graphql API.
             */
            var customerDS = _graphqlApi.AddDynamoDbDataSource("Customer", _customerTable);

            // Query Resolver to get all Customers
            customerDS.CreateResolver(new AppSync.BaseResolverProps
            {
                TypeName = "Query",
                FieldName = "getCustomers",
                RequestMappingTemplate = AppSync.MappingTemplate.DynamoDbScanTable(),
                ResponseMappingTemplate = AppSync.MappingTemplate.DynamoDbResultList()
            });

            // Query Resolver to get an individual Customer by their id
            customerDS.CreateResolver(new AppSync.BaseResolverProps
            {
                TypeName = "Query",
                FieldName = "getCustomer",
                RequestMappingTemplate = AppSync.MappingTemplate.DynamoDbGetItem("id", "id"),
                ResponseMappingTemplate = AppSync.MappingTemplate.DynamoDbResultItem()
            });

            // Mutation Resolver for adding a new Customer
            customerDS.CreateResolver(new AppSync.BaseResolverProps
            {
                TypeName = "Mutation",
                FieldName = "addCustomer",
                RequestMappingTemplate = AppSync.MappingTemplate.DynamoDbPutItem(
                    AppSync.PrimaryKey.Partition("id").Auto(),
                    AppSync.Values.Projecting("customer")
                ),
                ResponseMappingTemplate = AppSync.MappingTemplate.DynamoDbResultItem()
            });


            // Mutation Resolver for updating an exisiting Customer
            customerDS.CreateResolver(new AppSync.BaseResolverProps
            {
                TypeName = "Mutation",
                FieldName = "saveCustomer",
                RequestMappingTemplate = AppSync.MappingTemplate.DynamoDbPutItem(
                    AppSync.PrimaryKey.Partition("id").Is("id"),
                    AppSync.Values.Projecting("customer")
                ),
                ResponseMappingTemplate = AppSync.MappingTemplate.DynamoDbResultItem()
            });


            // Mutation resolver for creating a new customer along with their first order 
            customerDS.CreateResolver(new AppSync.BaseResolverProps
            {
                TypeName = "Mutation",
                FieldName = "saveCustomerWithFirstOrder",
                RequestMappingTemplate = AppSync.MappingTemplate.DynamoDbPutItem(
                    AppSync.PrimaryKey.Partition("order").Auto().Sort("customer").Is("customer.id"),
                    AppSync.Values.Projecting("order").Attribute("referral").Is("referral")
                ),
                ResponseMappingTemplate = AppSync.MappingTemplate.DynamoDbResultItem()
            });

            // Mutation Resolver for deleting an exisiting Customer
            customerDS.CreateResolver(new AppSync.BaseResolverProps
            {
                TypeName = "Mutation",
                FieldName = "removeCustomer",
                RequestMappingTemplate = AppSync.MappingTemplate.DynamoDbDeleteItem("id", "id"),
                ResponseMappingTemplate = AppSync.MappingTemplate.DynamoDbResultItem()
            });

            // defines an AWS Lambda resource
            _loyaltyLambda = new Lambda.Function(this, "LoyaltyLambdaHandler", new Lambda.FunctionProps
            {
                Runtime = Lambda.Runtime.NODEJS_12_X, // execution environment
                Code = Lambda.Code.FromAsset("lambda_fns"), // code loaded from the "lambda_fns" directory
                Handler = "loyalty.handler" // file is "loyalty", function is "handler"
            });
            
            /**
             * Add Loyalty Lambda as a Datasource for the Graphql API.
             */
            var loyaltyDS = _graphqlApi.AddLambdaDataSource("Loyalty", _loyaltyLambda);

            // Query Resolver to get all Customers
            loyaltyDS.CreateResolver(new AppSync.BaseResolverProps
            {
                TypeName = "Query",
                FieldName = "getLoyaltyLevel",
                RequestMappingTemplate = AppSync.MappingTemplate.LambdaRequest(),
                ResponseMappingTemplate = AppSync.MappingTemplate.LambdaResult()
            });

            // GraphQL API Endpoint
            new CfnOutput(this, "Endpoint", new CfnOutputProps
            {
                Value = _graphqlApi.GraphqlUrl
            });

            // API Key
            new CfnOutput(this, "API_Key", new CfnOutputProps
            {
                Value = _graphqlApi.ApiKey
            });

        }
    }
}
